Backport: Block template utils and rest templates controller

ntsekouras opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/56467

Backports changes from: 
-  https://github.com/WordPress/gutenberg/pull/41387
- https://github.com/WordPress/gutenberg/pull/41189
- https://github.com/WordPress/gutenberg/pull/42520
- https://github.com/WordPress/gutenberg/pull/42586


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ntsekouras committed:
Backport: Block template utils and rest templates controller

ntsekouras committed:
add `icon` to types controller

ntsekouras commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
```diff
@@ -678,4 +684,39 @@ protected function find_and_normalize_template_by_id( $templates, $id ) {
+	 * @covers WP_REST_Templates_Controller::get_template_fallback
+	 */
+	// TODO: check how to make this test work..
+	// https://github.com/WordPress/gutenberg/pull/42520/files#diff-5dcd7f0ba0e053098e93f6907913da73787fb7487e05d37e3249fb1c0ec434d6
```
How do you suggest to make these tests for core? In GB we have a different approach: https://github.com/WordPress/gutenberg/pull/42520/files#diff-5dcd7f0ba0e053098e93f6907913da73787fb7487e05d37e3249fb1c0ec434d6.

Should we create a new test theme or update an existing one or something else? 🤔 --cc @spacedmonkey

ntsekouras replied:
Used 'block-theme' eventually and tested with `slugs`.

ntsekouras committed:
update tests

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -525,6 +575,17 @@ protected function prepare_item_for_database( $request ) {
 			$changes->post_excerpt = $template->description;
 		}
 
+		if ( 'wp_template' === $this->post_type ) {
+			if ( isset( $request['is_wp_suggestion'] ) ) {
+				$changes->meta_input     = wp_parse_args(
+					array(
+						'is_wp_suggestion' => $request['is_wp_suggestion'],
+					),
+					$changes->meta_input = array()
+				);
+			}
+		}
```
Is there a unit test for this?

ntsekouras replied:
Added here: https://github.com/WordPress/wordpress-develop/pull/3221/commits/98878136a99e9c0e55376209f83833d9050e5e9e

ntsekouras committed:
add tests for `is_wp_suggestion`

ntsekouras commented:
~Do not merge before this GB PR: https://github.com/WordPress/gutenberg/pull/44085/~

---
edit: 

The GB PR has merged.

ockham reviewed (commented):
Thank you, @ntsekouras! This LGTM; however, we need to wait for Gutenberg 14.1 (which is the first release to contain WordPress/gutenberg#44085) to be released before merging this PR. Otherwise, a WP install running Core `trunk` and Gutenberg stable will fatal because of the duplicate `get_template_hierarchy` definition.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,65 @@ function wp_generate_block_templates_export_file() {
+ *
+ * @return array<string> The template hierarchy.
+ */
+function get_template_hierarchy( $slug, $is_custom = false, $template_prefix = '' ) {
```
Shouldn't this be prefixed with `wp_`? Why? 

- consistency
- to avoid any naming conflicts in the wild

hellofromtonya replied:
Oh I see, the other function names in that same file do not use the `wp_` prefix. Interesting. Okay, seems it's consistent within its context.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,65 @@ function wp_generate_block_templates_export_file() {
 	return $filename;
 }
+
+/**
+ * Helper function to get the Template Hierarchy for a given slug.
+ * We need to Handle special cases here like `front-page`, `singular` and `archive` templates.
+ *
+ * Noting that we always add `index` as the last fallback template.
+ *
+ * @since 6.1.0
+ *
+ * @param string  $slug            The template slug to be created.
+ * @param boolean $is_custom       Indicates if a template is custom or part of the template hierarchy.
+ * @param string  $template_prefix The template prefix for the created template. This is used to extract the main template type ex. in `taxonomy-books` we extract the `taxonomy`.
+ *
+ * @return array<string> The template hierarchy.
+ */
```
Some notes:
* avoid the word "we" in the codebase
* no space between the last `@param` and `@return` for consistency within Core
* Add `"Optional"` to each optional's parameter description
* Add the default value in each optional `@param`
* `array<string>` is not a PHP type. Use `string[]` instead.

```suggestion
/**
 * Helper function to get the Template Hierarchy for the given slug.
 * Handles special cases like `front-page`, `singular` and `archive` templates.
 *
 * Note: Always add `index` as the last fallback template.
 *
 * @since 6.1.0
 *
 * @param string  $slug           The template slug to be created.
 * @param boolean $is_custom      Optional. Indicates if a template is custom or
 *                                part of the template hierarchy. Default false.
 * @param string $template_prefix The template prefix for the created template.
 *                                Used to extract the main template type, e.g.
 *                                in `taxonomy-books` the `taxonomy` is extracted.
 *                                Default empty string.
 * @return string[] The template hierarchy.
 */
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -65,6 +66,33 @@ public function register_routes() {
+							'type'        => 'boolean',
+						),
+						'template_prefix' => array(
+							'description' => __( 'The template prefix for the created template. This is used to extract the main template type ex. in `taxonomy-books` we extract the `taxonomy`' ),
```
Avoid using the word "we" in the codebase.
```suggestion
							'description' => __( 'The template prefix for the created template. This is used to extract the main template type, e.g. in `taxonomy-books` extracts the `taxonomy`' ),
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -117,6 +145,28 @@ public function register_routes() {
 		);
 	}
 
+	/**
+	 * Returns the fallback template for a given slug.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param WP_REST_Request $request The request instance.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
```
No empty space between param and return.
```suggestion
	/**
	 * Returns the fallback template for the given slug.
	 *
	 * @since 6.1.0
	 *
	 * @param WP_REST_Request $request The request instance.
	 * @return WP_REST_Response|WP_Error
	 */
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -117,6 +145,28 @@ public function register_routes() {
+		if ( empty( $request['slug'] ) ) {
+			return new WP_Error(
+				'rest_invalid_param',
+				__( 'Invalid slug.', 'gutenberg' ),
```
Remove `'gutenberg'` text domain.
```suggestion
				__( 'Invalid slug.' ),
```

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -525,6 +575,17 @@ protected function prepare_item_for_database( $request ) {
 			$changes->post_excerpt = $template->description;
 		}
 
+		if ( 'wp_template' === $this->post_type ) {
+			if ( isset( $request['is_wp_suggestion'] ) ) {
```
These 2 conditionals can be combined as there's no other code within the first if block.

hellofromtonya commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -377,4 +377,160 @@ function test_wp_generate_block_templates_export_file() {
+	 *
+	 * @ticket 56467
+	 */
+	public function test_get_template_hierarchy() {
```
For each assertion in this test:

* Use `assertSame()` instead of `assertEquals`
* Add an assertion message (3rd arg) to describe it. Why? If an assertion fails, the rest will not run. By adding an assertion message, it helps to identify which assertion failed and why.

ntsekouras replied:
I updated all the test suggestions. Thanks for explaining the `why` for multiple assertions.

hellofromtonya commented on tests/phpunit/tests/rest-api/rest-post-types-controller.php:
```diff
@@ -174,6 +174,7 @@ public function test_get_item_schema() {
 		$this->assertArrayHasKey( 'rest_base', $properties );
 		$this->assertArrayHasKey( 'rest_namespace', $properties );
 		$this->assertArrayHasKey( 'visibility', $properties );
+		$this->assertArrayHasKey( 'icon', $properties );
```
Add an assertion message, 3rd arg.

hellofromtonya commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
```diff
@@ -678,4 +684,91 @@ protected function find_and_normalize_template_by_id( $templates, $id ) {
+			'is_custom'      => false,
+			'author'         => self::$admin_id,
+		);
+		$this->assertSame( $expected, $data );
+		// `is_wp_suggestion` false.
+		$body_params = wp_parse_args(
+			array(
+				'slug'             => 'page-hi',
+				'is_wp_suggestion' => false,
+			),
+			$body_params
+		);
+		$request->set_body_params( $body_params );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		unset( $data['_links'] );
+		unset( $data['wp_id'] );
+		$expected = wp_parse_args(
+			array(
+				'id'        => 'default//page-hi',
+				'slug'      => 'page-hi',
+				'is_custom' => true,
+			),
+			$expected
+		);
+		$this->assertSame( $expected, $data );
```
Add messages to each assertion.

hellofromtonya commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
```diff
@@ -678,4 +684,91 @@ protected function find_and_normalize_template_by_id( $templates, $id ) {
+		$request->set_param( 'is_custom', false );
+		$request->set_param( 'template_prefix', 'tag' );
+		$response = rest_get_server()->dispatch( $request );
+		$this->assertEquals( 'index', $response->get_data()->slug );
+		// Should fallback to `page.html` .
+		$request->set_param( 'slug', 'page-hello' );
+		$request->set_param( 'is_custom', false );
+		$request->set_param( 'template_prefix', 'page' );
+		$response = rest_get_server()->dispatch( $request );
+		$this->assertEquals( 'page', $response->get_data()->slug );
```
* use `assertSame()` instead of `assertEquals()`
* add a message to each assertion

hellofromtonya reviewed (changes requested):
First pass through the code. Most changes requested are for compliance to Core's coding / testing standards and/or for consistency.

hellofromtonya committed:
Update docblock for coding standards

hellofromtonya committed:
Remove 'gutenberg' text domain

hellofromtonya committed:
Improve docblock for coding standards

hellofromtonya committed:
Removes "we" from description

hellofromtonya committed:
Auto regenerated wp-api-generated.js qunit test fixture

hellofromtonya commented:
@spacedmonkey Are you happy with the REST code in this PR?

ntsekouras committed:
update tests

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:
```diff
@@ -221,6 +221,10 @@ public function prepare_item_for_response( $item, $request ) {
 			$data['slug'] = $post_type->name;
 		}
 
+		if ( in_array( 'icon', $fields, true ) ) {
```
```suggestion
		if ( rest_is_field_included( 'icon', $fields ) ) {
```

ntsekouras replied:
Should this be updated to all the other checks? If yes, it would be better to do in a separate PR, no?

spacedmonkey replied:
I don't mind doing this in this PR.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:
```diff
@@ -385,6 +390,12 @@ public function get_item_schema() {
+				'icon'           => array(
+					'description' => __( 'The icon for the post type.' ),
+					'type'        => 'string',
+					'context'     => array( 'view', 'edit', 'embed' ),
```
Does this need to be in the view context? The icon is not public right now?

ntsekouras replied:
What do you mean public? I think it would be good to have in `view` as it can be useful in some cases. Do you see any drawback being in `view` context?

spacedmonkey replied:
If you make a property as view it means it can't be viewed pubilically. Do we want to expose this data publically?

ntsekouras replied:
I see no harm in it, but if you have a strong opinion on this, I can remove it.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -117,6 +145,27 @@ public function register_routes() {
+		}
+		$hierarchy         = get_template_hierarchy( $request['slug'], $request['is_custom'], $request['template_prefix'] );
+		$fallback_template = resolve_block_template( $request['slug'], $hierarchy, '' );
+		return rest_ensure_response( $fallback_template );
```
```suggestion
		$response = $this->prepare_item_for_response( $fallback_template, $request );
		return rest_ensure_response( $response );
```

Please ensure that response is run through `prepare_item_for_response`, so the response is formatted correctly.

ntsekouras replied:
I'll update but this change will also need a GB PR(therefore a packages update) to retrieve the content from `content.raw`.

spacedmonkey replied:
Make the change and backport it. I can approve the change there as well. Change the properties / shape of response can / will cause problems down the road.

ntsekouras replied:
GB PR is here: https://github.com/WordPress/gutenberg/pull/44299 --cc @ockham

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -117,6 +145,27 @@ public function register_routes() {
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function get_template_fallback( $request ) {
+		if ( empty( $request['slug'] ) ) {
+			return new WP_Error(
+				'rest_invalid_param',
+				__( 'Invalid slug.' ),
+				array( 'status' => 400 )
+			);
+		}
```
This check should not be here. If you set `'required' => true,` to then this check should not be required. 

What if the slug was 0?

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -65,6 +66,33 @@ public function register_routes() {
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => array(
+						'slug'            => array(
+							'description' => __( 'The slug of the template to get the fallback for' ),
```
```suggestion
							'description' => __( 'The slug of the template to get the fallback for' ),
							'required' => true,
```

spacedmonkey replied:
There are now two required fields here.

ntsekouras replied:
Good catch, thanks!

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:
```diff
@@ -385,6 +390,12 @@ public function get_item_schema() {
 				),
+				'icon'           => array(
+					'description' => __( 'The icon for the post type.' ),
+					'type'        => 'string',
```
This value can be string or null?

ntsekouras replied:
It seems so if the post type hasn't declared a `menu_icon`. Should we update the `type` here? Do we need conditional adding to the response if the `menu_icon` is set? How do you handle this in general in core?

spacedmonkey replied:
You can have multiple types. See 
https://github.com/WordPress/wordpress-develop/blob/61f7e758952dabbe7e00cd2d9eeca2e7d82e4604/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2218

spacedmonkey commented on tests/phpunit/tests/block-template-utils.php:
```diff
@@ -377,4 +377,171 @@ function test_wp_generate_block_templates_export_file() {
+
+	/**
+	 * Helper function to get the Template Hierarchy for a given slug.
+	 *
```
Add covers comments here.

spacedmonkey commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
```diff
@@ -678,4 +684,91 @@ protected function find_and_normalize_template_by_id( $templates, $id ) {
 
+	/**
+	 * @ticket 56467
+	 * @covers WP_REST_Templates_Controller::create_item
```
Add covers.

ntsekouras replied:
I'm not sure what more to add besides the existing one.. Any suggestions?

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:
```diff
@@ -65,6 +66,33 @@ public function register_routes() {
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_template_fallback' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
```
Should this be `get_item_permissions_check`

ntsekouras committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

ntsekouras committed:
address feedback

ockham commented:
Thank you @spacedmonkey and @ntsekouras!

Unfortunately, it seems like [unit tests are failing](https://github.com/WordPress/wordpress-develop/actions/runs/3089920029/jobs/4998078712):

```
1) Tests_REST_WpRestTemplatesController::test_get_template_fallback
Trying to get property 'slug' of non-object

/var/www/tests/phpunit/tests/rest-api/wpRestTemplatesController.php:766
phpvfscomposer:///var/www/vendor/phpunit/phpunit/phpunit:97
/var/www/vendor/bin/phpunit:118
```

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:
```diff
@@ -221,6 +221,10 @@ public function prepare_item_for_response( $item, $request ) {
 		}
 
+		if ( in_array( 'icon', $fields, true ) ) {
+			$data['icon'] = $post_type->menu_icon;
```
If this value is null, should it maybe default to something?

ntsekouras replied:
I think it's okay to be null since it's not required and the consumers can decide what to do with it. For example in php core for post types we assign some default icon if it's in the `admin menu`, but in the site editor we assign different default icon..

ntsekouras committed:
feedback

ntsekouras committed:
update tests and use `rest_is_field_included`

ntsekouras committed:
remove duplicate required

hellofromtonya commented:
Currently reviewing. When ready, will prep the commit.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+	$template_hierarchy = array( $slug );
+	// Most default templates don't have `$template_prefix` assigned.
+	if ( $template_prefix ) {
+		list($type) = explode( '-', $template_prefix );
```
```suggestion
		list( $type ) = explode( '-', $template_prefix );
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+	// Most default templates don't have `$template_prefix` assigned.
+	if ( $template_prefix ) {
+		list($type) = explode( '-', $template_prefix );
+		// We need these checks because we always add the `$slug` above.
```
Remove "we"

```suggestion
		// These checks are needed because the `$slug` above is always added.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+	if ( 'front-page' === $slug ) {
+		return array( 'front-page', 'home', 'index' );
+	}
+	$template_hierarchy = array( $slug );
```
```suggestion

	$template_hierarchy = array( $slug );

```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+			$template_hierarchy[] = $type;
+		}
+	}
+	// Handle `archive` template.
```
```suggestion

	// Handle `archive` template.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+	) {
+		$template_hierarchy[] = 'singular';
+	}
+	$template_hierarchy[] = 'index';
```
```suggestion

	$template_hierarchy[] = 'index';

```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+	if ( 'attachment' === $slug ) {
+		$template_hierarchy[] = 'single';
+	}
+	// Handle `singular` template.
```
```suggestion

	// Handle `singular` template.
```

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1012,3 +1014,68 @@ function wp_generate_block_templates_export_file() {
+ * @param string  $slug           The template slug to be created.
+ * @param boolean $is_custom      Optional. Indicates if a template is custom or
+ *                                part of the template hierarchy. Default false.
+ * @param string $template_prefix The template prefix for the created template.
```
```suggestion
 * @param string $template_prefix Optional. The template prefix for the created template.
```

hellofromtonya committed:
Code review updates for get_template_hierarchy()

hellofromtonya committed:
Test updates get_template_hierarchy().

* Moved tests to its own test file.
* Moved individual assertions to a data provider.

hellofromtonya committed:
Updates to REST API tests

hellofromtonya commented:
Rebased the branch on top of the latest `trunk`.

hellofromtonya committed:
Reapply changeset 54184.

After rebasing on top of trunk, the PR reverted some
of the changes in wp-includes/block-template-utils.php
that were added in changest 54184.

This commit reapplies the code from that changeset.

It also cleans up in one test.

hellofromtonya reviewed (approved):
LTGM 👍 Ready for commit, which I'm prepping next.

hellofromtonya commented:
Note: The npm CI job failure is unrelated to this PR.

hellofromtonya commented on src/wp-includes/block-template-utils.php:
```diff
@@ -1023,3 +1026,74 @@ function wp_generate_block_templates_export_file() {
 }
+
+/**
+ * Helper function to get the Template Hierarchy for the given slug.
+ * Handles special cases like `front-page`, `singular` and `archive` templates.
```
```suggestion
 * Gets the template hierarchy for the given template slug to be created.
 *
```

hellofromtonya committed:
Improve DocBlock.

hellofromtonya commented:
Committed via changeset https://core.trac.wordpress.org/changeset/54269. Thank you everyone for your contributions!
